_posts/2018-09-21-first-class-resolved-file.md

```diff
@@ -0,0 +1,194 @@
+
+To detect those kind of problems, we've added a new entry `output_tree_hash`
+to the dict describing a repository. For example, the entry for
+`com_google_protobuf` in the resolved file now looks as follows.
+
+```
+resolved = [
+    ...
+    {
+        "original_rule_class": "@bazel_tools//tools/build_defs/repo:git.bzl%git_repository",
+        "original_attributes": {
+            "name": "com_google_protobuf",
+            "remote": "https://github.com/google/protobuf",
+            "branch": "master"
+        },
+        "repositories": [
+            {
+                "rule_class": "@bazel_tools//tools/build_defs/repo:git.bzl%git_repository",
+                "output_tree_hash": "a776ce4f591327c6b23d88d367d6208a88af6ad889e08f7b86a0edfc76fcfd96",
+                "attributes": {
+                    "remote": "https://github.com/google/protobuf",
+                    "commit": "a6e1cc7e328c45a0cb9856c530c8f6cd23314163",
+                    "shallow_since": "2018-09-17",
+                    "init_submodules": False,
+                    "verbose": False,
+                    "strip_prefix": "",
+                    "patches": [],
+                    "patch_tool": "patch",
+                    "patch_args": [
+                        "-p0"
+                    ],
+                    "patch_cmds": [],
+                    "name": "com_google_protobuf"
+                }
+            }
+        ]
+    }
+
+]
+```
+
+This new `output_tree_hash` entry is a [hash of the
+directory](https://github.com/bazelbuild/proposals/blob/master/designs/2018-07-13-repository-hashing.md)
+generated by the repository rule. It does include names and contents of all
+files, as well as the executability bit. However, information that is likely to
+be different between various users and won't affect most builds (like owner of
+the files or the modification time) is ignored. Also for symlinks to files
+outside the repository, the content of the file is hashed, rather than the link
+path itself; so the link generated by a `build_file` argument is not a problem.
+
+The resolved file to take the hashes from can be specifed with the
+`--experimental_repository_hash_file` option. Of course, not for all types
+of "external repositories" we even expect reproducible content. For example
+the `cc_autoconf` rule is specifically designed to detect the local `C++`
+toolchain, which might well differ from machine to machine. So, you can
+specify with the `--experimental_verify_repository_rules` option for which
+rule classes to verify the generated repository. For example,
+`bazel build --experimental_repository_hash_file=resolved.bzl
+--experimental_verify_repository_rules=@bazel_tools//tools/build_defs/repo:git.bzl%git_repository
+...` will verify the hashes of all git repositories, but not do any
+verification for repositories generated by other rules.
+
+If a mismatch is found, e.g., because you added the not-so-hermetic
```

Comment:
Replace "e.g." with "For example". (Google's standard practice is to never use "e.g." or "i.e." because non-English speakers may not know what they stand for.)